File: leetcode/42TrappingRainWater.cpp
// LeetCode Problem 42: Trapping Rain Water
// https://leetcode.com/problems/trapping-rain-water/
#include <vector>
using namespace std;
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;
        int ans = 0;
        int left_max = 0;
        int right_max = 0;

        while(left < right){
            if (height[left] < height[right]) {
                if (height[left] >= left_max) {
                    left_max = height[left];
                } else {
                    ans += left_max - height[left];
                }
                ++left;
            }else {
                if (height[right] >= right_max) {
                    right_max = height[right];
                } else {
                    ans += right_max - height[right];
                }
                --right;
            }
        }
        return ans;
    }
};

// Time Complexity: O(n)
// Space Complexity: O(1)

/*
This solution uses a two-pointer approach to calculate the amount of trapped rain water.
dry run:
height = [0,1,0,2,1,0,1,3,2,1,2,1]
left | right | left_max | right_max | height[left] | height[right] | ans
-----------------------------------------------------------
    0   |  11   |    0     |     0      |      0       |       1       |  0
    1   |  11   |    1     |     0      |      1       |       1       |  0
    1   |  10   |    1     |     1      |      1       |       2       |  0
    2   |  10   |    1     |     2      |      0       |       2       |  1
    3   |  10   |    2     |     2      |      2       |       2       |  1
    3   |  9    |    2     |     2      |      2       |       1       |  2
    4   |  9    |    2     |     2      |      1       |       1       |  3
    5   |  9    |    2     |     2      |      0       |       1       |  5
    6   |  9    |    2     |     2      |      1       |       1       |  5
    7   |  9    |    3     |     2      |      3       |       1       |  5
    7   | 8     |    3     |     2      |      3       |       2       | 5
    End:  
return ans = 6

[0,1,0,2,1,0,1,3,2,1,2,1]

init:
left=0 right=11 left_max=0 right_max=0 ans=0

loop 1:
height[left]=0 < height[right]=1
0 >= 0 → left_max=0
left=1

loop 2:
height[left]=1 < 1 ✘
1 >= 0 → right_max=1
right=10

loop 3:
height[left]=1 < 2
1 >= 0 → left_max=1
left=2

loop 4:
height[left]=0 < 2
0 < 1 → ans += 1-0 = 1 (ans=1)
left=3

loop 5:
height[left]=2 < 2 ✘
2 >= 1 → right_max=2
right=9

loop 6:
height[left]=2 < 1 ✘
1 < 2 → ans += 2-1 = 1 (ans=2)
right=8

loop 7:
height[left]=2 < 2 ✘
2 >= 2 → right_max=2
right=7

loop 8:
height[left]=2 < 3
2 >= 1 → left_max=2
left=4

loop 9:
height[left]=1 < 3
1 < 2 → ans += 2-1 = 1 (ans=3)
left=5

loop 10:
height[left]=0 < 3
0 < 2 → ans += 2-0 = 2 (ans=5)
left=6

loop 11:
height[left]=1 < 3
1 < 2 → ans += 2-1 = 1 (ans=6)
left=7

stop (left == right)

return ans = 6

*/

File: leetcode/python/42TrappingRainWater.py

# leetcode 42. Trapping Rain Water
from typing import List
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height)-1
        left_max, right_max = 0, 0

        water = 0

        while l < r:
            if height[l] < height[r]:
                if height[l] >= left_max:
                    left_max = height[l]
                else:
                    water += left_max - height[l]
                l += 1
            else:
                if height[r] >= right_max:
                    right_max = height[r]
                else:
                    water += right_max - height[r]
                r -=1
        return water

'''
height = [0,1,0,2,1,0,1,3,2,1,2,1]

l = 0, r = 11
left_max = 0, right_max = 0
water = 0


Step 1:
height[l]=0, height[r]=1 → process left
0 >= 0 → left_max = 0
l = 1

Step 2:
height[l]=1, height[r]=1 → process right
1 >= 0 → right_max = 1
r = 10

Step 3:
height[l]=1, height[r]=2 → process left
1 >= 0 → left_max = 1
l = 2

Step 4:
height[l]=0, height[r]=2 → process left
0 < 1 → water += 1
water = 1
l = 3

Step 5:
height[l]=2, height[r]=2 → process right
2 >= 1 → right_max = 2
r = 9

Step 6:
height[l]=2, height[r]=1 → process right
1 < 2 → water += 1
water = 2
r = 8

Step 7:
height[l]=2, height[r]=2 → process right
2 >= 2 → right_max = 2
r = 7

Step 8:
height[l]=2, height[r]=3 → process left
2 >= 1 → left_max = 2
l = 4

Step 9:
height[l]=1, height[r]=3 → process left
1 < 2 → water += 1
water = 3
l = 5

Step 10:
height[l]=0, height[r]=3 → process left
0 < 2 → water += 2
water = 5
l = 6

Step 11:
height[l]=1, height[r]=3 → process left
1 < 2 → water += 1
water = 6
l = 7


Stop: l == r

Final water = 6
'''